File: m/rls/fimod.r
file	fimod - file operations
include	rid:rider
  If Vms
include	<unixio>			; has fileno()
  Elif Dos				;
#undef __STDC__				;
include	<stdio.h>			; has fileno()
  End					;
include	rid:fidef			; also includes <stdio.h>
include	rid:imdef			;
include	rid:stdef			;
include	rid:medef			;
include	rid:osdef			;

;	o  Filename defaults
;	o  Fully qualified filespecs
;	o  Truenames
;	o  Local translation of filenames
;	o  Messages (and default messages) with local filenames
;	o  Improve on standard C's lack of concern for read/write errors

	E_OpnFil := "E-Error opening file [%s]"
	E_CreFil := "E-Error creating file [%s]"
	E_PrcFil := "E-Error processing file [%s]"
	E_CloFil := "E-Error closing file [%s]"
	E_DelFil := "E-Error deleting file [%s]"
	E_RenFil := "E-Error renaming file [%s]"
	E_ExsFil := "E-Missing file [%s]"
	E_MisFil := "E-File already exists [%s]"

  type	fiTinf
  is	Vlen : short			; filespec slot size
	Pspc : * char			; file spec
	Pfil : * FILE			; some implementations only
;	Verr : int			; errno
;	Vcod : int			; local error
  end

	fiPrep	: * imTrep = &im_rep	; report function
	fiAinf	: [fiCmax] fiTinf own = {0} ; info array

	fi_reg	: (*FILE, *char) int own; register file
	fi_drg	: (*FILE) void own	; deregister file
	fi_inf	: (*FILE) * fiTinf own	; get file info slot
code	fi_opn - open file

;	Add logic to handle sys$input/stdin etc
;	if *mod eq 'w'			; a new file
;	   fil->Pfil = fopen (spc, mod, "RFM=STM") ; create stream file
;
;	"r"	read, input
;	"w"	write, create
;	"rw"	input and output
;	"b"	binary

  func	fi_opn
	spc : * char			; filespec
	mod : * char			; fopen mode
	msg : * char			; the message
	()  : * FILE			; file or null
  is	fil : * FILE			;
	loc : [fiLspc] char		; local copy
	if !st_sam (spc, "con")
	   fi_loc (spc, loc)		; localize the filespec
	else
If Wnt
	   st_cop ("conin$", loc) if st_fnd ("r", mod)
	   st_cop ("conout$", loc) otherwise
End
	end
;PUT("opn=[%s] ", loc)
	if (fil = fopen (loc, mod)) ne	; open it
	   fi_reg (fil, loc)		; register it
	.. reply fil			; opened it
	reply null if msg eq 		; they do it
	if *msg eq			; default message
	   if st_mem ('r', mod) eq	; not read
	   && st_mem ('w', mod) ne	; and explicit write
	      msg = E_CreFil		; is a create error
	   else				; anything else, including
	.. .. msg = E_OpnFil		; errors is an open
	fi_rep (msg, loc, msg)		; tell them
	reply null			;
  end

code	fi_clo - close file

;	o  Ignores NULL files
;	o  Reports either processing or close error 

  func	fi_clo
	fil : * FILE
	msg : * char			; the message
	()  : int			; fail => errors detected
  is	sta : int			;
	fine if fil eq <>		; null files have no errors
	sta = fi_chk (fil, msg)		; check it
	fi_drg (fil)			; deregister it first
	reply sta if fclose (fil) ne EOF; closed without error
	fail if sta eq			; already failed
	fail fi_rep (msg, fi_spc (fil), E_CloFil); report it
  end
code	fi_del - delete file

  func	fi_del
	spc : * char
	msg : * char			; the message
	()  : int			; fine/fail
  is	loc : [fiLspc] char		;
	fi_loc (spc, loc)		; make a copy
	fine if remove (loc) eq		; dump it
	fine if os_del (loc, <>)	; delete it
	fail fi_rep (msg, spc, E_DelFil);
  end

code	fi_ren - rename file

  func	fi_ren
	old : * char
	new : * char
	msg : * char
  is	src : [(fiLspc*2)+2] char	; space for both names in a message
	dst : [fiLspc] char		;
	fi_loc (old, src)		; localize them
	fi_loc (new, dst)		;
	fine if rename (src, dst) eq	;
	st_app ("] [", src)		; join the names
	st_app (dst, src)		; Error renaming [src] [dst]
	fail fi_rep (msg, src, E_RenFil); report it
  end
code	fi_exs - check file exists

  func	fi_exs
	spc : * char
	msg : * char			; exists message
  is	fil : * FILE
	fil = fi_opn (spc, "rb", <>) 	; no such file
	fine fi_clo (fil, <>) if fil	; no problems
	fail fi_rep (msg, spc, E_ExsFil); handle messages
  end

code	fi_mis - check file missing

  func	fi_mis
	spc : * char
	msg : * char			; exists message
  is	fil : * FILE
	fil = fi_opn (spc, "rb", <>) 	; test file
	fine if !fil			; no such - great
	fi_clo (fil, <>)		; close it first
	fail fi_rep (msg, spc, E_MisFil); handle messages
  end

code	fi_chk - check file for errors

  func	fi_chk
	fil : * FILE
	msg : * char
	()  : int
  is	spc : * char			; the filespec
	fine if ferror (fil) eq		; no errors
	spc = fi_spc (fil)		; get the filespec
	fail fi_rep (msg, spc, E_PrcFil); handle messages
  end

code	fi_rep - report file error

  func	fi_rep
	msg : * char
	spc : * char
	def : * char			; default message
  	()  : int			; always fine
  is	fine if msg eq <>		; no message required
	msg = def if *msg eq		; want default message
	(*fiPrep)(msg, spc)		; report it
	fine
  end
code	fi_spc - workout filespec

  func	fi_spc
	fil : * FILE			;
	()  : * char			; filespec or null
  is	inf : * fiTinf			;
	reply "(null)" if fil eq <>	; common error
	inf = fi_inf (fil)		; get the information slot 
	reply "(invalid)" if null	; invalid file number
	reply inf->Pspc if inf->Pspc ne	; known filespec
	reply "(stdin)" if fil eq stdin	;
	reply "(stdout)" if fil eq stdout ;
	reply "(stderr)" if fil eq stderr ;
	reply "(unknown)"		;
  end

code	fi_reg - register spec

  func	fi_reg
	fil : * FILE
	spc : * char
	()  : int			; fine/fail
  is	inf : * fiTinf			;
	len : int = st_len (spc) + 1	;
	inf = fi_inf (fil)		; get the slot
	fail if null			; out of touch
	inf->Pfil = fil			; registered
	if inf->Vlen lt len		;
	   inf->Vlen = (len + 31) & ~(31) ; avoid fragmentation
	.. me_alp (<**void>&inf->Pspc, that) ; ZTC fails
	st_cop (spc, inf->Pspc)		; save it
	fine				;
  end

code	fi_drg - deregister file

;	File is deregisted by clearing inf->Pfil
;	This variable will be used on systems lacking fileno()

  proc	fi_drg
	fil : * FILE
  is	inf : * fiTinf
	inf = fi_inf (fil)		; get the slot
	exit if null			; is none
	inf->Pfil = <>			; deregister
  end

code	fi_inf - get info entry address

  func	fi_inf
	fil : * FILE			;
	()  : * fiTinf			; spec or null
  is	idx : unsigned			;
	idx = fileno (fil)		; get the filenumber
	reply <> if idx ge fiCmax	; too high
	reply &fiAinf[idx]		; the entry
  end
end file
code	fi_sav - setup file context

;	fiPspc		most recent filename from open, delete, rename
;			fully qualified and localized
;
;	fiPmsg		default message for most recent error
;			null if most recent operation succeeded
;
;	fiPctx		pointer to user defined context 
;
;	fiPrep		pointer to default report routine
;			null for muted messages
;			defaults to im_rep

  proc	fi_sav
	sav : * fiTsav			; the save area
	rep : * () int			; user report function
	ctx : * void			; user report context (if any)
	()  : * fiTsav			;
  is	if sav eq <>			; dynamic
	   sav = me_acc (#fiTsav)	; allocate & clear it
	.. ++sav->Vdyn			; remember its dynamic
	sav->Pmsg = fiPmsg		; save it
	sav->Pspc = fiPspc		;
	sav->Prep = fiPrep		;
	sav->Pctx = riPctx		;
	fiPmsg = <>			; no message
	fiPspc = <>			; or spec
	fiPrep = rep			;
	fiPrep = &im_rep if null	; default to im_rep
	fiPctx = ctx			; user context
	reply sav			;
  end

code	fi_res - restore context

  proc	fi_res
	sav : * fiTsav
  is	fiPmsg = sav->Pmsg
	fiPspc = sav->Pspc
	fiPrep = sav->Prep
	fiPctx = sav->Pctx
	me_dlc (sav) if sav->Vdyn	; deallocate dynamic block
  end


File: m/rls/rlb/fimod.c
/* file -  fimod - file operations */
#include "m:\rid\rider.h"
#if Vms
#include <unixio>
#elif  Dos
#undef __STDC__
#include <stdio.h>
#endif 
#include "m:\rid\fidef.h"
#include "m:\rid\imdef.h"
#include "m:\rid\stdef.h"
#include "m:\rid\medef.h"
#define E_OpnFil  "E-Error opening file [%s]"
#define E_CreFil  "E-Error creating file [%s]"
#define E_PrcFil  "E-Error processing file [%s]"
#define E_CloFil  "E-Error closing file [%s]"
#define E_DelFil  "E-Error deleting file [%s]"
#define E_RenFil  "E-Error renaming file [%s]"
#define E_ExsFil  "E-Missing file [%s]"
#define E_MisFil  "E-File already exists [%s]"
#define fiTinf struct fiTinf_t 
struct fiTinf_t
{ short Vlen ;
  char *Pspc ;
  FILE *Pfil ;
   };
imTrep *fiPrep = &im_rep;
static fiTinf fiAinf [fiCmax]= {0};
static int fi_reg (FILE *,char *);
static void fi_drg (FILE *);
static fiTinf *fi_inf (FILE *);
/* code -  fi_opn - open file */
FILE *fi_opn(
char *spc ,
char *mod ,
char *msg )
{ FILE *fil ;
  char loc [fiLspc];
  if ( !st_sam (spc, "con")) {
    fi_loc (spc, loc);
    } else {
#if Wnt
    if ( st_fnd ("r", mod)) {st_cop ("conin$", loc) ;} else {
      st_cop ("conout$", loc) ; }
#endif 
  } 
  if ( (fil = fopen (loc, mod)) != 0) {
    fi_reg (fil, loc);
    return ( fil); }
  if( msg == 0)return ( NULL );
  if ( *msg == 0) {
    if (( st_mem ('r', mod) == 0)
    &&(st_mem ('w', mod) != 0)) {
      msg = E_CreFil;
      } else {
      msg = E_OpnFil; } }
  fi_rep (msg, loc, msg);
  return ( NULL);
} 
/* code -  fi_clo - close file */
int fi_clo(
FILE *fil ,
char *msg )
{ int sta ;
  if( fil == NULL)return 1;
  sta = fi_chk (fil, msg);
  fi_drg (fil);
  if( fclose (fil) != EOF)return ( sta );
  if( sta == 0)return 0;
   fi_rep (msg, fi_spc (fil), E_CloFil);return 0;
} 
/* code -  fi_del - delete file */
int fi_del(
char *spc ,
char *msg )
{ char loc [fiLspc];
  fi_loc (spc, loc);
  if( remove (loc) == 0)return 1;
   fi_rep (msg, spc, E_DelFil);return 0;
} 
/* code -  fi_ren - rename file */
fi_ren(
char *old ,
char *new ,
char *msg )
{ char src [(fiLspc*2)+2];
  char dst [fiLspc];
  fi_loc (old, src);
  fi_loc (new, dst);
  if( rename (src, dst) == 0)return 1;
  st_app ("] [", src);
  st_app (dst, src);
   fi_rep (msg, src, E_RenFil);return 0;
} 
/* code -  fi_exs - check file exists */
fi_exs(
char *spc ,
char *msg )
{ FILE *fil ;
  fil = fi_opn (spc, "rb", NULL);
  if( fil){ fi_clo (fil, NULL) ; return 1;}
   fi_rep (msg, spc, E_ExsFil);return 0;
} 
/* code -  fi_mis - check file missing */
fi_mis(
char *spc ,
char *msg )
{ FILE *fil ;
  fil = fi_opn (spc, "rb", NULL);
  if( !fil)return 1;
  fi_clo (fil, NULL);
   fi_rep (msg, spc, E_MisFil);return 0;
} 
/* code -  fi_chk - check file for errors */
int fi_chk(
FILE *fil ,
char *msg )
{ char *spc ;
  if( ferror (fil) == 0)return 1;
  spc = fi_spc (fil);
   fi_rep (msg, spc, E_PrcFil);return 0;
} 
/* code -  fi_rep - report file error */
int fi_rep(
char *msg ,
char *spc ,
char *def )
{ if( msg == NULL)return 1;
  if ( *msg == 0) {msg = def ;}
  (*fiPrep)(msg, spc);
  return 1;
} 
/* code -  fi_spc - workout filespec */
char *fi_spc(
FILE *fil )
{ fiTinf *inf ;
  if( fil == NULL)return ( "(null)" );
  if( (inf = fi_inf (fil)) == NULL)return ( "(invalid)" );
  if( inf->Pspc != 0)return ( inf->Pspc );
  if( fil == stdin)return ( "(stdin)" );
  if( fil == stdout)return ( "(stdout)" );
  if( fil == stderr)return ( "(stderr)" );
  return ( "(unknown)");
} 
/* code -  fi_reg - register spec */
int fi_reg(
FILE *fil ,
char *spc )
{ fiTinf *inf ;
  int len = st_len (spc) + 1;
  if( (inf = fi_inf (fil)) == NULL)return 0;
  inf->Pfil = fil;
  if ( inf->Vlen < len) {
    me_alp ((void **)&inf->Pspc, (inf->Vlen = (len + 31) & ~(31))); }
  st_cop (spc, inf->Pspc);
  return 1;
} 
/* code -  fi_drg - deregister file */
void fi_drg(
FILE *fil )
{ fiTinf *inf ;
  if( (inf = fi_inf (fil)) == NULL)return;
  inf->Pfil = NULL;
} 
/* code -  fi_inf - get info entry address */
fiTinf *fi_inf(
FILE *fil )
{ unsigned idx ;
  idx = fileno (fil);
  if( idx >= fiCmax)return ( NULL );
  return ( &fiAinf[idx]);
} 
